src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -0,0 +1,746 @@
+            return null;
+        }
+        I item = items.get(name);
+        if (item == null) {
+            return null;
+        }
+        if (!item.hasPermission(Item.READ)) {
+            if (item.hasPermission(Item.DISCOVER)) {
+                throw new AccessDeniedException("Please log in to access " + name);
+            }
+            return null;
+        }
+        return item;
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public void onRenamed(TopLevelItem item, String oldName, String newName) throws IOException {
+        items.remove(oldName);
+        items.put(newName, (I) item);
+        // For compatibility with old views:
+        for (View v : views) {
+            v.onJobRenamed(item, oldName, newName);
+        }
+        save();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public void onDeleted(TopLevelItem item) throws IOException {
+        ItemListener.fireOnDeleted(item);
+        items.remove(item.getName());
+        // For compatibility with old views:
+        for (View v : views) {
+            v.onJobRenamed(item, item.getName(), null);
+        }
+        save();
+    }
+
+    @Override
+    protected void performDelete() throws IOException, InterruptedException {
+        // delete individual items first
+        // (disregard whether they would be deletable in isolation)
+        SecurityContext orig = ACL.impersonate(ACL.SYSTEM);
+        try {
+            for (Item i : new ArrayList<Item>(items.values())) {
+                try {
+                    i.delete();
+                } catch (AbortException e) {
+                    throw (AbortException) new AbortException(
+                            "Failed to delete " + i.getFullDisplayName() + " : " + e.getMessage()).initCause(e);
+                } catch (IOException e) {
+                    throw new IOException("Failed to delete " + i.getFullDisplayName(), e);
+                }
+            }
+        } finally {
+            SecurityContextHolder.setContext(orig);
+        }
+        super.performDelete();
+    }
+
+    @Override
+    public synchronized void save() throws IOException {
+        if (BulkChange.contains(this)) {
+            return;
+        }
+        super.save();
+        // TODO should this not just be done in AbstractItem?
+        ItemListener.fireOnUpdated(this);
+    }
+
+    /**
+     * Renames this item container.
+     */
+    @Override
+    public void renameTo(String newName) throws IOException {
+        super.renameTo(newName);
+    }
+
+    @Override
+    public synchronized void doSubmitDescription(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
+        getPrimaryView().doSubmitDescription(req, rsp);
+    }
+
+    public void doConfigSubmit(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, Descriptor.FormException {
+        checkPermission(CONFIGURE);
+
+        req.setCharacterEncoding("UTF-8");
```

Comment:
I think a comment here would be useful as this looks a bit strange (setting the char encoding of the request).